go-controller/pkg/ovn/controller/apbroute/external_controller_namespace.go

```diff
@@ -1,90 +1,73 @@
-// * Calculate the difference between current and expected policies and proceed to remove the gateway IPs from the policies that are no longer applicable to this namespace
-// * Calculate the difference between the expected and current ones to determine the new policies to be applied and proceed to apply them.
-// * Update the cache info with the new list of policies, as well as the static and dynamic gateway IPs derived from executing the previous logic.
-func (m *externalPolicyManager) processUpdateNamespace(namespaceName string, currentPolicies, newPolicies sets.Set[string], cacheInfo *namespaceInfo) error {
-
-	// some differences apply, let's figure out if previous policies have been removed first
-	policiesNotValid := currentPolicies.Difference(newPolicies)
-	// iterate through the policies that no longer apply to this namespace
-	for policyName := range policiesNotValid {
-		err := m.removePolicyFromNamespaceWithName(namespaceName, policyName, cacheInfo)
+func (m *externalPolicyManager) syncNamespace(namespace *v1.Namespace, routeQueue workqueue.RateLimitingInterface) error {
+	klog.V(5).Infof("APB processing namespace: %s", namespace)
+
+	keysToBeQueued := sets.New[string]()
+
+	// Get a copy of all policies at this time and see if they include this namespace
+	policyKeys := m.routePolicySyncCache.GetKeys()
+
+	for _, policyName := range policyKeys {
+		err := m.routePolicySyncCache.DoWithLock(policyName,
+			func(key string) error {
+				ri, found := m.routePolicySyncCache.Load(policyName)
+				if found {
+					targetNsList, err := m.listNamespacesBySelector(&ri.policy.Spec.From.NamespaceSelector)
+					if err != nil {
+						return err
+					}
+					for _, targetNS := range targetNsList {
+						if targetNS.Name == namespace.Name {
+							keysToBeQueued.Insert(policyName)
+						}
+					}
+					for _, hop := range ri.policy.Spec.NextHops.DynamicHops {
+						gwNs, err := m.listNamespacesBySelector(hop.NamespaceSelector)
+						if err != nil {
+							return err
+						}
+						for _, ns := range gwNs {
+							if ns.Name == namespace.Name {
+								keysToBeQueued.Insert(policyName)
+							}
+						}
+					}
+				}
+				return nil
+			})
 		if err != nil {
 			return err
 		}
 	}
 
-	// policies that now apply to this namespace
-	newPoliciesDiff := newPolicies.Difference(currentPolicies)
-	for policyName := range newPoliciesDiff {
-		policy, found, markedForDeletion := m.getRoutePolicyFromCache(policyName)
-		if !found {
-			return fmt.Errorf("failed to find external route policy %s in cache", policyName)
-		}
-		if markedForDeletion {
-			klog.Infof("Skipping route policy %s as it has been marked for deletion", policyName)
-			continue
-		}
-		err := m.applyPolicyToNamespace(namespaceName, &policy, cacheInfo)
-		if err != nil {
-			return err
+	// Check if this namespace is being tracked by policy in its namespace cache
+	cacheInfo, found := m.getNamespaceInfoFromCache(namespace.Name)
```

Comment:
I agree this race could happen. I dont think passing the old namespace will work because then we wouldnt have the new namespace labels to check if the policy applies (first part of the function).

We could lock the namespaceInfoCache key for the namespace in policy handler (step 1) as soon as we list from kapi to find the namespaces. That would prevent step 3 from happening until the policy was done processing the namespace. Although then we would be locking all the namespaces up front in the cache in the policy handler, that could create deadlock across policy handlers.

What about if we lock the policy key in 1? That would prevent the first loop:
```
	for _, policyName := range policyKeys {
		err := m.routePolicySyncCache.DoWithLock(policyName,
```

from completing until the policy has finished being added. At that point we know the namespaceInfo cache is up to date?

Yes, we should make sure the dynamic hops are accounted for.

Reply:
I meant pass both new and old namespace to check which labels has changed, but we certainly can consider other options.
With the current implementation, policy is added to the cache at the very end of its handling, so locking won't help, but maybe if we add policy to the cache in the beginning of its handling, that can work

Reply:
What if we use the `onUpdateNamespace` to detect the diff in policies based on the nsSelectors for target and dynamic hops from the informer directly and queue the policies impacted?